### PR TITLE
Bug 1609081 - Provide full server side search in Alerts view

### PR DESCRIPTION
**Pull request by ionutgoldan**

### CODE REVIEW

File: ui/perfherder/alerts/AlertsView.jsx [hunk 1]
```diff
@@ -98,20 +122,48 @@ class AlertsView extends React.Component {
...
+    const params = {};
+    for (const [filterName, filterValue] of Object.entries(filters)) {
+      Object.assign(params, this.mapParamFromFilter(filterName, filterValue));
+    }
+    return params;
+  };
+
+  // TODO: find a more elegant solution for this
+  mapParamFromFilter = (filterName, filterValue) => {
+    if (filterName === 'framework') {
```
Comment (ionutgoldan) [inline ui/perfherder/alerts/AlertsView.jsx]: I think a basic `switch` statement would suffice, by improving readibility. Not sure it's worth adding a key-lambda dictionary.

File: tests/ui/perfherder/alerts_test.jsx [hunk 2]
```diff
@@ -664,28 +464,21 @@ describe('"My alerts" checkbox\'s display behaviors', () => {
...
   test('Not displayed in Alerts view (detailed mode)', async () => {
     const { queryByText } = alertsViewControls({
-      isListingAlertSummaries: false,
+      isListMode: false,
     }); // as Django detailed view mode
 
     expect(queryByText('My alerts')).not.toBeInTheDocument();
   });
 
-  test('Not displayed in Alerts view (detailed mode), even when param is missing', async () => {
```
Comment (octavian-negru) [inline tests/ui/perfherder/alerts_test.jsx]: Just a question: why this test case was deleted?

File: treeherder/webapp/api/performance_data.py [hunk 3]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
     page_size = 10
 
 
+class PerformanceAlertSummaryFilter(django_filters.FilterSet):
```
Comment (octavian-negru) [inline treeherder/webapp/api/performance_data.py]: This is a little piece of backend mastery. 
It is beyond my knowledge and I would suggest to add a more Django-experienced developer to ensure that the `PerformanceAlertSummaryFilter` is OK.

Comment (airimovici) [reply treeherder/webapp/api/performance_data.py]: @ionutgoldan it would be a good idea to see also the compiled version of these queries, just to be have a good picture of how it is transformed

File: ui/perfherder/alerts/AlertsView.jsx [hunk 4]
```diff
@@ -70,6 +77,19 @@ class AlertsView extends React.Component {
       this.setState({ totalPages: this.generatePages(count) });
     }
 
+    if (
+      validated.hideImprovements !== prevProps.validated.hideImprovements ||
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsView.jsx]: I would give a name to `validated.hideImprovements !== prevProps.validated.hideImprovements`, such as `hideImprovementsChanged`.
The same for `hideDwnToInv`.

File: ui/perfherder/alerts/AlertsView.jsx [hunk 5]
```diff
@@ -98,20 +122,48 @@ class AlertsView extends React.Component {
...
-    const { updateParams } = this.props.validated;
-    const { frameworkOptions } = this.state;
-    const framework = frameworkOptions.find(item => item.name === selection);
-    updateParams({ framework: framework.id });
-    this.setState({ framework, bugTemplate: null }, () =>
-      this.fetchAlertSummaries(),
-    );
+  getDefaultFilterText = () => {
+    const { filterText } = this.props.validated;
+    return filterText === undefined || filterText === null ? '' : filterText;
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsView.jsx]: I would rewrite as `!filterText ? '' : filterText;`, since `undefined`, `null` and `''` are falsy.

File: ui/perfherder/alerts/AlertsView.jsx (hunk 1 above)
Comment (octavian-negru) [reply ui/perfherder/alerts/AlertsView.jsx]: Also, a `switch` would replace the `if`s with `else if`s, thus making the check faster.

File: ui/perfherder/alerts/AlertsView.jsx [hunk 6]
```diff
@@ -278,12 +333,20 @@ class AlertsView extends React.Component {
             </Alert>
           )}
           <AlertsViewControls
-            dropdownOptions={id ? [] : alertDropdowns}
+            isListMode={this.isListMode()}
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsView.jsx]: Suggestion: all filters can be grouped in a single object.
Benefit: less verbosity.

File: ui/perfherder/alerts/AlertsView.jsx [hunk 7]
```diff
@@ -172,6 +228,21 @@ class AlertsView extends React.Component {
     let updates = { loading: false };
     const params = this.composeParams(id, page, framework, status);
 
+    if (this.isListMode()) {
+      if (filterText !== '') {
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsView.jsx]: What happens if `filterText` is `undefined` or `null`?
Maybe using `if (filterText)` would suffice as a check.

File: ui/perfherder/alerts/AlertsView.jsx [hunk 8]
```diff
@@ -172,6 +228,21 @@ class AlertsView extends React.Component {
...
+        params.filter_text = filterText;
+      }
+      if (hideImprovements) {
+        params.hide_improvements = hideImprovements;
+      }
+      if (hideDownstream) {
+        // params.hide_downstream = hideDownstream;
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsView.jsx]: I think that the `// params.hide_downstream` comment is something which got forgotten as a comment.

File: ui/perfherder/alerts/AlertsViewControls.jsx [hunk 9]
```diff
@@ -127,15 +122,22 @@ AlertsViewControls.propTypes = {
...
-  isListingAlertSummaries: PropTypes.bool,
-  dropdownOptions: PropTypes.arrayOf(PropTypes.shape({})),
+  isListMode: PropTypes.bool.isRequired,
+  filterText: PropTypes.string.isRequired,
+  hideImprovements: PropTypes.bool.isRequired,
+  hideDownstream: PropTypes.bool.isRequired,
+  hideAssignedToOthers: PropTypes.bool.isRequired,
+  framework: PropTypes.shape({}).isRequired,
+  status: PropTypes.string.isRequired,
+  // TODO: rename to ~updateFilters
```
Comment (octavian-negru) [inline ui/perfherder/alerts/AlertsViewControls.jsx]: This TODO is needed?

File: ui/shared/InputFilter.jsx [hunk 10]
```diff
@@ -33,15 +34,32 @@ export default class InputFilter extends React.Component {
...
+    };
+
+    if (updateOnEnter) {
+      inputProps.onKeyDown = this.userActionListener;
+    }
+
     return (
       <InputGroup>
         <Input
+          {...inputProps}
```
Comment (octavian-negru) [inline ui/shared/InputFilter.jsx]: Suggestion: please expand this `{...inputProps}` since this would make the code more explicit.

File: tests/ui/perfherder/alerts_test.jsx (hunk 2 above)
Comment (airimovici) [reply tests/ui/perfherder/alerts_test.jsx]: This does not apply anymore, because I'm passing a prop from AlertsView to know for sure if we are in `ListMode` or `DetailMode` (`isListMode`) and that prop is required, eliminating the edge case the test was covering.

We still have this test above: `Not displayed in Alerts view (detailed mode)`

File: ui/perfherder/alerts/AlertsViewControls.jsx (hunk 9 above)
Comment (airimovici) [reply ui/perfherder/alerts/AlertsViewControls.jsx]: still thinking of a better name for that function :)

File: treeherder/webapp/api/performance_data.py [hunk 11]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
...
+    status = django_filters.NumberFilter(field_name='status')
+    framework = django_filters.NumberFilter(field_name='framework')
+    repository = django_filters.NumberFilter(field_name='repository')
+    alerts__series_signature = django_filters.NumberFilter(field_name='alerts__series_signature')
+    filter_text = django_filters.CharFilter(method='text_to_filter')
+    hide_improvements = django_filters.BooleanFilter(method='_hide_improvements')
+    hide_downstream = django_filters.BooleanFilter(method='_hide_downstream')
+    with_assignee = django_filters.CharFilter(method='_with_assignee')
+
+    def text_to_filter(self, queryset, name, value):
```
Comment (ionutgoldan) [inline treeherder/webapp/api/performance_data.py]: Equivalent raw SQL query, assuming the filter text is "startup linux stylo":
```sql
SELECT `performance_alert_summary`.`id`,
       `performance_alert_summary`.`repository_id`,
       `performance_alert_summary`.`framework_id`,
       `performance_alert_summary`.`prev_push_id`,
       `performance_alert_summary`.`push_id`,
       `performance_alert_summary`.`manually_created`,
       `performance_alert_summary`.`notes`,
       `performance_alert_summary`.`assignee_id`,
       `performance_alert_summary`.`created`,
       `performance_alert_summary`.`first_triaged`,
       `performance_alert_summary`.`last_updated`,
       `performance_alert_summary`.`status`,
       `performance_alert_summary`.`bug_number`,
       `performance_alert_summary`.`bug_updated`,
       `performance_alert_summary`.`issue_tracker_id`,
       `repository`.`id`,
       `repository`.`repository_group_id`,
       `repository`.`name`,
       `repository`.`dvcs_type`,
       `repository`.`url`,
       `repository`.`branch`,
       `repository`.`codebase`,
       `repository`.`description`,
       `repository`.`active_status`,
       `repository`.`performance_alerts_enabled`,
       `repository`.`expire_performance_data`,
       `repository`.`is_try_repo`,
       `repository`.`tc_root_url`,
       `push`.`id`,
       `push`.`repository_id`,
       `push`.`revision`,
       `push`.`author`,
       `push`.`time`
FROM `performance_alert_summary`
         INNER JOIN `repository` ON (`performance_alert_summary`.`repository_id` = `repository`.`id`)
         INNER JOIN `push` ON (`performance_alert_summary`.`push_id` = `push`.`id`)
WHERE (`repository`.`active_status` = 'active' AND `performance_alert_summary`.`status` = 0 AND
       `performance_alert_summary`.`framework_id` = 1 AND `performance_alert_summary`.`id` IN (SELECT DISTINCT U0.`id`
                                                                                               FROM `performance_alert_summary` U0
                                                                                                        INNER JOIN `repository` U1 ON (U0.`repository_id` = U1.`id`)
                                                                                                        LEFT OUTER JOIN `performance_alert` U3 ON (U0.`id` = U3.`summary_id`)
                                                                                                        LEFT OUTER JOIN `performance_signature` U4 ON (U3.`series_signature_id` = U4.`id`)
                                                                                                        LEFT OUTER JOIN `machine_platform` U5 ON (U4.`platform_id` = U5.`id`)
                                                                                                        INNER JOIN `push` U6 ON (U0.`push_id` = U6.`id`)
                                                                                               WHERE (U1.`active_status` = 'active' AND
                                                                                                      U0.`status` = 0 AND
                                                                                                      U0.`framework_id` = 1 AND
                                                                                                      CONCAT_WS('',
                                                                                                                U4.`suite`,
                                                                                                                CONCAT_WS(
                                                                                                                        '',
                                                                                                                        CONCAT_WS(
                                                                                                                                '',
                                                                                                                                U4.`test`,
                                                                                                                                CONCAT_WS(
                                                                                                                                        '',
                                                                                                                                        CONCAT_WS(
                                                                                                                                                '',
                                                                                                                                                U5.`platform`,
                                                                                                                                                CONCAT_WS(
                                                                                                                                                        '',
                                                                                                                                                        CONCAT_WS(
                                                                                                                                                                '',
                                                                                                                                                                U4.`extra_options`,
                                                                                                                                                                CONCAT_WS(
                                                                                                                                                                        '',
                                                                                                                                                                        CONCAT_WS('', U0.`bug_number`, CONCAT_WS('', U6.`revision`)))))))))) LIKE
                                                                                                      BINARY 'startup' AND
                                                                                                      CONCAT_WS('',
                                                                                                                U4.`suite`,
                                                                                                                CONCAT_WS(
                                                                                                                        '',
                                                                                                                        CONCAT_WS(
                                                                                                                                '',
                                                                                                                                U4.`test`,
                                                                                                                                CONCAT_WS(
                                                                                                                                        '',
                                                                                                                                        CONCAT_WS(
                                                                                                                                                '',
                                                                                                                                                U5.`platform`,
                                                                                                                                                CONCAT_WS(
                                                                                                                                                        '',
                                                                                                                                                        CONCAT_WS(
                                                                                                                                                                '',
                                                                                                                                                                U4.`extra_options`,
                                                                                                                                                                CONCAT_WS(
                                                                                                                                                                        '',
                                                                                                                                                                        CONCAT_WS('', U0.`bug_number`, CONCAT_WS('', U6.`revision`)))))))))) LIKE
                                                                                                      BINARY '%linux%' AND
                                                                                                      CONCAT_WS('',
                                                                                                                U4.`suite`,
                                                                                                                CONCAT_WS(
                                                                                                                        '',
                                                                                                                        CONCAT_WS(
                                                                                                                                '',
                                                                                                                                U4.`test`,
                                                                                                                                CONCAT_WS(
                                                                                                                                        '',
                                                                                                                                        CONCAT_WS(
                                                                                                                                                '',
                                                                                                                                                U5.`platform`,
                                                                                                                                                CONCAT_WS(
                                                                                                                                                        '',
                                                                                                                                                        CONCAT_WS(
                                                                                                                                                                '',
                                                                                                                                                                U4.`extra_options`,
                                                                                                                                                                CONCAT_WS(
                                                                                                                                                                        '',
                                                                                                                                                                        CONCAT_WS('', U0.`bug_number`, CONCAT_WS('', U6.`revision`)))))))))) LIKE
                                                                                                      BINARY '%stylo%')))
```

File: treeherder/webapp/api/performance_data.py [hunk 12]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
...
+                'bug_number', sep,
+                'push__revision',
+                output_field=CharField()))
+                                      .filter(*contains_all_words)
+                                      .values('id')
+                                      .distinct())
+
+        return queryset.filter(id__in=Subquery(filtered_summaries))
+
+    def _hide_improvements(self, queryset, name, value):
```
Comment (ionutgoldan) [inline treeherder/webapp/api/performance_data.py]: Equivalent raw SQL query, when `Hide improvements` is checked:
```sql
SELECT `performance_alert_summary`.`id`,
       `performance_alert_summary`.`repository_id`,
       `performance_alert_summary`.`framework_id`,
       `performance_alert_summary`.`prev_push_id`,
       `performance_alert_summary`.`push_id`,
       `performance_alert_summary`.`manually_created`,
       `performance_alert_summary`.`notes`,
       `performance_alert_summary`.`assignee_id`,
       `performance_alert_summary`.`created`,
       `performance_alert_summary`.`first_triaged`,
       `performance_alert_summary`.`last_updated`,
       `performance_alert_summary`.`status`,
       `performance_alert_summary`.`bug_number`,
       `performance_alert_summary`.`bug_updated`,
       `performance_alert_summary`.`issue_tracker_id`,
       COUNT(`performance_alert`.`is_regression`) AS `total_regressions`,
       `repository`.`id`,
       `repository`.`repository_group_id`,
       `repository`.`name`,
       `repository`.`dvcs_type`,
       `repository`.`url`,
       `repository`.`branch`,
       `repository`.`codebase`,
       `repository`.`description`,
       `repository`.`active_status`,
       `repository`.`performance_alerts_enabled`,
       `repository`.`expire_performance_data`,
       `repository`.`is_try_repo`,
       `repository`.`tc_root_url`,
       `push`.`id`,
       `push`.`repository_id`,
       `push`.`revision`,
       `push`.`author`,
       `push`.`time`
FROM `performance_alert_summary`
         INNER JOIN `repository` ON (`performance_alert_summary`.`repository_id` = `repository`.`id`)
         LEFT OUTER JOIN `performance_alert` ON (`performance_alert_summary`.`id` = `performance_alert`.`summary_id`)
         INNER JOIN `performance_alert` T5 ON (`performance_alert_summary`.`id` = T5.`summary_id`)
         INNER JOIN `push` ON (`performance_alert_summary`.`push_id` = `push`.`id`)
WHERE (`repository`.`active_status` = 'active' AND `performance_alert_summary`.`status` = 0 AND
       `performance_alert_summary`.`framework_id` = 1 AND T5.`is_regression` = True)
GROUP BY `performance_alert_summary`.`id`
HAVING COUNT(`performance_alert`.`is_regression`) >= 1
ORDER BY NULL
```

File: treeherder/webapp/api/performance_data.py [hunk 13]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
...
+                .annotate(total_regressions=Count('alerts__is_regression'))
+                .filter(alerts__is_regression=True, total_regressions__gte=1))
+
+    def _hide_related_and_invalid(self, queryset, name, value):
+        return (queryset.exclude(
+            state__in=[PerformanceAlertSummary.DOWNSTREAM,
+                       PerformanceAlertSummary.REASSIGNED,
+                       PerformanceAlertSummary.INVALID]))
+
+    def _with_assignee(self, queryset, name, value):
```
Comment (ionutgoldan) [inline treeherder/webapp/api/performance_data.py]: Equivalent raw SQL query, when `My alerts` is checked *(and I'm logged in)*:
```sql
SELECT `performance_alert_summary`.`id`,
       `performance_alert_summary`.`repository_id`,
       `performance_alert_summary`.`framework_id`,
       `performance_alert_summary`.`prev_push_id`,
       `performance_alert_summary`.`push_id`,
       `performance_alert_summary`.`manually_created`,
       `performance_alert_summary`.`notes`,
       `performance_alert_summary`.`assignee_id`,
       `performance_alert_summary`.`created`,
       `performance_alert_summary`.`first_triaged`,
       `performance_alert_summary`.`last_updated`,
       `performance_alert_summary`.`status`,
       `performance_alert_summary`.`bug_number`,
       `performance_alert_summary`.`bug_updated`,
       `performance_alert_summary`.`issue_tracker_id`,
       `repository`.`id`,
       `repository`.`repository_group_id`,
       `repository`.`name`,
       `repository`.`dvcs_type`,
       `repository`.`url`,
       `repository`.`branch`,
       `repository`.`codebase`,
       `repository`.`description`,
       `repository`.`active_status`,
       `repository`.`performance_alerts_enabled`,
       `repository`.`expire_performance_data`,
       `repository`.`is_try_repo`,
       `repository`.`tc_root_url`,
       `push`.`id`,
       `push`.`repository_id`,
       `push`.`revision`,
       `push`.`author`,
       `push`.`time`
FROM `performance_alert_summary`
         INNER JOIN `repository` ON (`performance_alert_summary`.`repository_id` = `repository`.`id`)
         INNER JOIN `auth_user` ON (`performance_alert_summary`.`assignee_id` = `auth_user`.`id`)
         INNER JOIN `push` ON (`performance_alert_summary`.`push_id` = `push`.`id`)
WHERE (`repository`.`active_status` = 'active' AND `performance_alert_summary`.`status` = 0 AND
       `performance_alert_summary`.`framework_id` = 1 AND `auth_user`.`username` = 'mozilla-ldap/igoldan@mozilla.com')
```

File: treeherder/webapp/api/performance_data.py [hunk 14]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
...
+                                      .distinct())
+
+        return queryset.filter(id__in=Subquery(filtered_summaries))
+
+    def _hide_improvements(self, queryset, name, value):
+        return (queryset
+                .annotate(total_regressions=Count('alerts__is_regression'))
+                .filter(alerts__is_regression=True, total_regressions__gte=1))
+
+    def _hide_related_and_invalid(self, queryset, name, value):
```
Comment (ionutgoldan) [inline treeherder/webapp/api/performance_data.py]: Equivalent raw SQL query, when `Hide downstream / reassigned to / invalid` is checked:
```sql
SELECT `performance_alert_summary`.`id`,
       `performance_alert_summary`.`repository_id`,
       `performance_alert_summary`.`framework_id`,
       `performance_alert_summary`.`prev_push_id`,
       `performance_alert_summary`.`push_id`,
       `performance_alert_summary`.`manually_created`,
       `performance_alert_summary`.`notes`,
       `performance_alert_summary`.`assignee_id`,
       `performance_alert_summary`.`created`,
       `performance_alert_summary`.`first_triaged`,
       `performance_alert_summary`.`last_updated`,
       `performance_alert_summary`.`status`,
       `performance_alert_summary`.`bug_number`,
       `performance_alert_summary`.`bug_updated`,
       `performance_alert_summary`.`issue_tracker_id`,
       `repository`.`id`,
       `repository`.`repository_group_id`,
       `repository`.`name`,
       `repository`.`dvcs_type`,
       `repository`.`url`,
       `repository`.`branch`,
       `repository`.`codebase`,
       `repository`.`description`,
       `repository`.`active_status`,
       `repository`.`performance_alerts_enabled`,
       `repository`.`expire_performance_data`,
       `repository`.`is_try_repo`,
       `repository`.`tc_root_url`,
       `push`.`id`,
       `push`.`repository_id`,
       `push`.`revision`,
       `push`.`author`,
       `push`.`time`
FROM `performance_alert_summary`
         INNER JOIN `repository` ON (`performance_alert_summary`.`repository_id` = `repository`.`id`)
         INNER JOIN `push` ON (`performance_alert_summary`.`push_id` = `push`.`id`)
WHERE (`repository`.`active_status` = 'active' AND `performance_alert_summary`.`status` = 0 AND
       `performance_alert_summary`.`framework_id` = 1 AND NOT (`performance_alert_summary`.`status` IN (1, 2, 3)))
```

File: ui/perfherder/alerts/AlertsView.jsx [hunk 15]
```diff
@@ -70,6 +77,19 @@ class AlertsView extends React.Component {
       this.setState({ totalPages: this.generatePages(count) });
     }
 
+    if (
```
[the remark this answers is not in the document]
Comment (airimovici) [reply ui/perfherder/alerts/AlertsView.jsx]: We are doing this to sychronize the changes from the url to the filters

File: ui/perfherder/alerts/AlertsView.jsx (hunk 5 above)
Comment (airimovici) [reply ui/perfherder/alerts/AlertsView.jsx]: I was afraid I might run into other falsy values, that should not be considered falsy

Comment (octavian-negru) [reply ui/perfherder/alerts/AlertsView.jsx]: Fair enough. 
If `filterText` is 0, this code should treat 0 as truthy. 
So you are interested in protecting against `undefined` or `null`.

**Comment by ionutgoldan** [comment]
@sarah-clements could you review this PR?

**Comment by ionutgoldan** [comment]
@camd could you please review the backend code?

**Comment by ionutgoldan** [comment]
@armenzg I noticed you're using `treeherder-prototype` for testing [bug 1608427](https://bugzilla.mozilla.org/show_bug.cgi?id=1608427)'s PR. We need to let the perf sheriffs test this PR on that environment as well, to have their feedback.
Could you lend us `treeherder-prototype` for next Monday *(Feb 24)*? Thanks!

**Comment by ionutgoldan** [comment]
I've also added the perf sheriffs as reviewers, in advance, so they're more aware of this PR.

**Comment by armenzg** [comment]
I've released prototype.

File: treeherder/webapp/api/performance_data.py (hunk 12 above)
Comment (sarah-clements) [reply treeherder/webapp/api/performance_data.py]: Did you use the django shell-plus or debug toolbar to get the SQL statements?

File: ui/perfherder/alerts/AlertsView.jsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (sarah-clements) [reply ui/perfherder/alerts/AlertsView.jsx]: `prevValitated` contains a typo.

File: treeherder/webapp/api/performance_data.py [hunk 16]
```diff
@@ -267,24 +274,78 @@ class AlertSummaryPagination(pagination.PageNumberPagination):
...
+                'alerts__series_signature__platform__platform', sep,
+                'alerts__series_signature__extra_options', sep,
+                'bug_number', sep,
+                'push__revision',
+                output_field=CharField()))
+                                      .filter(*contains_all_words)
+                                      .values('id')
+                                      .distinct())
+
+        return queryset.filter(id__in=Subquery(filtered_summaries))
```
Comment (sarah-clements) [inline treeherder/webapp/api/performance_data.py]: Are these results paginated?

Comment (ionutgoldan) [reply treeherder/webapp/api/performance_data.py]: Yes, they are automatically paginated by our custom Django-based code. More precisely, [here](https://github.com/mozilla/treeherder/blob/6e09f58bf48b8be2ffdfaa10462a4480ed7069e8/treeherder/webapp/api/performance_data.py#L289).

Comment (sarah-clements) [reply treeherder/webapp/api/performance_data.py]: I can't remember if I've mentioned this before, but you might find the [CustomPaginator](https://github.com/mozilla/treeherder/blob/6e09f58bf48b8be2ffdfaa10462a4480ed7069e8/treeherder/webapp/api/pagination.py#L13) useful if you end up profiling this API for performance and see issues. In the past, the count method was performing a full table scan when returning paginated results.
